utils/vptr/tests/basic.cc

```diff
@@ -0,0 +1,212 @@
+    ASSERT_NE(ptrA, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(ptrA));
+    ASSERT_EQ(pMap.count(), 1u);
+
+    void *ptrB = SYCLmalloc(10 * sizeof(int), pMap);
+
+    ASSERT_NE(ptrB, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(ptrA));
+    ASSERT_EQ(pMap.count(), 2u);
+
+    // Obtain the buffer
+    // Note that the scope of this buffer ends when the buffer
+    // is freed
+    try {
+      auto b2 = pMap.get_buffer(ptrB);
+      auto b1 = pMap.get_buffer(ptrA);
+
+#ifdef __COMPUTECPP__
+      ASSERT_NE(b2.get_impl().get(), b1.get_impl().get());
+#endif  // __COMPUTECPP__
+
+      cl::sycl::queue q([&](cl::sycl::exception_list e) {
+        std::cout << "Error " << std::endl;
+      });
+
+      q.submit([&b1, &b2](cl::sycl::handler &h) {
+        auto accB1 = b1.get_access<sycl_acc_rw>(h);
+        auto accB2 = b2.get_access<sycl_acc_rw>(h);
+        h.single_task<class foo2>([=]() {
+          cl::sycl::codeplay::get_device_ptr_as<int>(accB1)[0] = 1;
+          cl::sycl::codeplay::get_device_ptr_as<int>(accB2)[0] = 2;
+        });
+      });
+
+      q.wait_and_throw();
+      // Only way of reading the value is using a host accessor
+      {
+        auto hostAccA = b1.get_access<sycl_acc_rw, sycl_acc_host>();
+        ASSERT_EQ(cl::sycl::codeplay::get_host_ptr_as<int>(hostAccA)[0], 1);
+      }
+      {
+        auto hostAccB = b2.get_access<sycl_acc_rw, sycl_acc_host>();
+        ASSERT_EQ(cl::sycl::codeplay::get_host_ptr_as<int>(hostAccB)[0], 2);
+      }
+    } catch (std::out_of_range e) {
+      FAIL();
+    }
+
+    ASSERT_EQ(pMap.count(), 2u);
+    SYCLfree(ptrA, pMap);
+    ASSERT_EQ(pMap.count(), 1u);
+    SYCLfree(ptrB, pMap);
+    ASSERT_EQ(pMap.count(), 0u);
+  }
+}
+
+TEST(pointer_mapper, reuse_ptr) {
+  PointerMapper pMap;
+  {
+    ASSERT_EQ(pMap.count(), 0u);
+
+    // First we insert a large buffer
+    void *initial = SYCLmalloc(100 * sizeof(int), pMap);
+    ASSERT_NE(initial, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(initial));
+    ASSERT_EQ(pMap.count(), 1u);
+
+    // Now we insert a small one, that we'll be reused
+    void *reused = SYCLmalloc(10 * sizeof(int), pMap);
+    ASSERT_NE(reused, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(reused));
+    ASSERT_EQ(pMap.count(), 2u);
+
+    // Another large buffer
+    void *end = SYCLmalloc(100 * sizeof(int), pMap);
+    ASSERT_NE(end, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(end));
+    ASSERT_EQ(pMap.count(), 3u);
+
+    // We free the intermediate one
+    SYCLfree(reused, pMap);
+    ASSERT_EQ(pMap.count(), 2u);
+
+    void *shouldBeTheSame = SYCLmalloc(10 * sizeof(int), pMap);
+    ASSERT_NE(shouldBeTheSame, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(shouldBeTheSame));
+    ASSERT_EQ(pMap.count(), 3u);
+    ASSERT_EQ(shouldBeTheSame, reused);
+  }
+}
+
+TEST(pointer_mapper, multiple_alloc_free) {
+  PointerMapper pMap;
+  size_t numAllocations = (1 << 9);
+  size_t maxAllocSize = 1251;
```

Comment:
Is this an arbitrary magic number, or is it related to the max amount of memory that *could* be allocated, given that many allocations?